### ISSUE TITLE
Announcement: .NET Core 3.0 concludes the .NET Framework API porting project

### ISSUE THREAD

**Issue by terrajobst** [dotnet/corefx#41769]
We started in .NET Core 1.0 with a very minimal API set that only included \~18K of the .NET Framework APIs. With [.NET Standard 2.0](https://devblogs.microsoft.com/dotnet/announcing-net-standard-2-0/), we tried to make it much more viable to share code between .NET Framework, .NET Core, and Xamarin which resulted in approximately 38K .NET Frameworks APIs being available in .NET Core 2.0. We also built the [Windows Compatibility Pack](https://devblogs.microsoft.com/dotnet/announcing-the-windows-compatibility-pack-for-net-core/) which made another 21K .NET Framework APIs available to .NET Core, resulting in almost 60K additional APIs. And in .NET Core 3.0 we added WPF and WinForms, which increased the number of .NET Framework APIs ported to .NET Core to over 120k, which is more than half of all .NET Framework APIs.

It’s also worth pointing out that we added about 62K APIs to .NET Core that don’t exist in .NET Framework. If we compare their total number of APIs, .NET Core has about 80% of the API surface of .NET Framework.

<p align="center">
<img src="https://user-images.githubusercontent.com/5169960/66777114-f8db7c80-ee7c-11e9-9161-acfe1c491586.png" width="600" align="center">
</p>

We announced that the [future of .NET](https://devblogs.microsoft.com/dotnet/net-core-is-the-future-of-net/) will be based on .NET Core. And at Build 2019, [Scott Hunter stated](https://www.youtube.com/watch?v=ZlO1utbB2GQ&t=54m20s) that AppDomains, remoting, Web Forms, WCF server, and Windows Workflow won’t be ported to .NET Core.

With .NET Core 3.0, we’re at the point where we’ve ported all technologies that are required for modern workloads, be that desktop apps, mobile apps, console apps, web sites, or cloud services. That’s not to say that we don’t have any gaps or opportunities for new technologies, but we generally believe we won’t be finding them in the .NET Framework code base anymore. Moving forward, we’re focusing our resources on incorporating new technologies.

Simultaneously, we’re looking into releasing more of the [.NET Framework code base under the MIT license](https://github.com/microsoft/referencesource) on GitHub to allow the community to create OSS projects for technologies we’re not intending to bring to .NET Core. For example, there already are community projects for [CoreWF](https://github.com/UiPath/corewf) and [CoreWCF](https://github.com/CoreWCF/CoreWCF).

We’d like to thank everyone who filed issues with requests for APIs being ported. Those issues allowed us to prioritize and close the gaps that prevented people from porting to .NET Core.

But since we generally no longer plan to bring existing technologies from .NET Framework to .NET Core we’ll be closing all issues that are [labeled with port-to-core](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+label%3Aport-to-core).

**Comment by raffaeler** [comment]
Any chance to see the original codebase for WCF server published (not maintained)? An archived repository with all the sources would be perfect for forking.

**Comment by terrajobst** [comment]
@raffaeler 

> Any chance to see the original codebase for WCF server published (not maintained)? An archived repository with all the sources would be perfect for forking.

Yes. AFAIK we've already exposed all of WCF in the [MIT licensed .NET Framework code](https://github.com/microsoft/referencesource). That's why I wrote this:

> Simultaneously, we’re looking into releasing more of the [.NET Framework code base under the MIT license](https://github.com/microsoft/referencesource) on GitHub to allow the community to create OSS projects for technologies we’re not intending to bring to .NET Core. For example, there already are community projects for [CoreWF](https://github.com/UiPath/corewf) and [CoreWCF](https://github.com/CoreWCF/CoreWCF).

**Comment by raffaeler** [comment]
Thank you @terrajobst, I already knew about CoreWF and CoreWCF, but some customers have strong policies about the origin where the sources are published, that is why I asked.
While practically WCF is not "hidden", for legal reasons, it would be easier for them to see the sources published (even if unmaintained) in a GitHub repository.

Thanks

**Comment by ltrzesniewski** [comment]
Does that mean we won't get [LambdaExpression.CompileToMethod](https://docs.microsoft.com/en-us/dotnet/api/system.linq.expressions.lambdaexpression.compiletomethod) in .NET Core? I had hoped it was simply delayed because of assembly dependency issues (it's currently [hidden behind a feature flag](https://github.com/dotnet/corefx/blob/f4344b0d3b45fc7c8f41e5c909ddcb2752bea0fa/src/System.Linq.Expressions/src/System/Linq/Expressions/LambdaExpression.cs#L135-L149))...

**Comment by terrajobst** [comment]
> While practically WCF is not "hidden", for legal reasons it would be easier for them to see the sources published (even if unmaintained) in a GitHub repository.

Right, but the published WCF sources won't be usable by themselves (because the reference source doesn't include any of our internal build infrastructure). So whatever community is working on it, will have to copy the relevant sources into a new repo and a build environment for it (project files, packaging etc). That's basically what CoreWCF already is.

**Comment by terrajobst** [comment]
> Does that mean we won't get [LambdaExpression.CompileToMethod](https://docs.microsoft.com/en-us/dotnet/api/system.linq.expressions.lambdaexpression.compiletomethod) in .NET Core? I had hoped it was simply delayed because of assembly dependency issues (it's currently [hidden behind a feature flag](https://github.com/dotnet/corefx/blob/f4344b0d3b45fc7c8f41e5c909ddcb2752bea0fa/src/System.Linq.Expressions/src/System/Linq/Expressions/LambdaExpression.cs#L135-L149))...

This announcement is about technologies (e.g. expression trees or RefEmit itself). I see minor convenience APIs like `CompileToMethod` still as fair game, because they can be thought of as an evolution of .NET Core. So it's really more about the motivation: we wouldn't add this API just because it existed in .NET Framework. But by the same token, we wouldn't not do it, just because it already existed in .NET Framework either.  So feel free to file an issue for that.

**Comment by ltrzesniewski** [comment]
Thanks, that's good to know! It's already logged under #16287 BTW.

**Comment by willdean** [comment]
Is `CSharpCodeProvider` ever going to happen in Core?  The CodeDom API arrived somewhere in the 2.x timeframe, but it's a runtime fail even under 3.0 - are we going to need to re-write for some kind of explicitly Roslyn scripting?

**Comment by terrajobst** [comment]
`CSharpCodeProvider` is supported on .NET Core, but only the portions that deal with CodeDOM creation. You can't compile because this would require a compiler and the .NET Core runtime doesn't include a compiler (i.e. Roslyn). However, we might be able to do some gymnastics, like reflection, where the app can pull that in. Still tracked under #12180.

**Comment by kolappannathan** [comment]
Any plans for local speech synthesis if you are not porting System.Speech?

**Comment by willdean** [comment]
@terrajobst The problem with #12180 is that it doesn't really help us choose between A. changing our existing code to support Core and B. merely waiting for Core to support our existing code.  It's not really a very active issue (14 months since any substantive activity) and is neither a 'no we're never going to make this work' nor a 'yes it's on the roadmap for 5.0 (or whenever)'.

I was only really trying to find out if this latest announcement offered any additional clarity on that.  

As you say we can "just change the code" to use Roslyn explicitly, it would just be nice if we didn't need to.

**Comment by terrajobst** [comment]
@kolappannathan 

> Any plans for local speech synthesis, if you are not porting System.Speech?

I'll ask for details.

@willdean 

> The problem with #12180 is that it doesn't really help us choose between A. changing our existing code to support Core and B. merely waiting for Core to support our existing code.

I think the idea of #12180 is to make the existing CodeDOM API work. The only additional gesture for the developer might be to add some kind of NuGet package that glues CodeDOM and Roslyn together.

> It's not really a very active issue (14 months since any substantive activity) and is neither a 'no we're never going to make this work' nor a 'yes it's on the roadmap for 5.0 (or whenever)'.

Usage of this APIs fairly low (~0 on NuGet, ~1% in API Port telemetry). We've prioritized APIs with more usage. Also, this API is mostly convenience; the workaround is to invoke Roslyn on the generated source files.

> As you say we can "just change the code" to use Roslyn explicitly, it would just be nice if we didn't need to.

Right, and there is nothing wrong with wanting this. It's just that (1) it's a low usage API and (2) it has a reasonable workaround and (3) supporting requires more work than a straight port.

**Comment by dotMorten** [comment]
Could you also opensource System.Devices? This bit isn't on reference source.

**Comment by coxp** [comment]
Really disappointed with MS on this one. I don't expect a full port of WCF server-side but surely having the ability to create SOAP endpoints is still necessary for any enterprise projects on any platform. Telling us to use the .NET Framework or a proxy service just doesn't cut it. Yes, there are attempts at open-source projects (SoapCore and CoreWCF) but neither work and are incompatible with clients created against .NET framework SOAP services. I'd love to see the telemetry on how WCF was used in general as I'd imagine most services are just simple basic HTTP endpoints and that doesn't seem like much work to port.

**Comment by jhaygood86** [comment]
> Yes, there are attempts at open-source projects (SoapCore and CoreWCF) but neither work and are incompatible with clients created against .NET framework SOAP services.

@coxp -- Do you have any details about what makes SoapCore incompatible? My employer did a porting effort that required minimal code changes to go from .NET Framework WCF to SoapCore (actually required LESS code), and didn't require ANY client side or contract changes (we literally just changed the host name from our web server farm running IIS to our app server running since we self-hosted the SoapCore version via Kestrel not using IIS). Our contract is also stored in an internal nuget package we reference from servers and client, and we referenced the same package in both the WCF/IIS version and the SoapCore/Kestrel version. I looked through the SoapCore repo and didn't see any issues filed.

**Comment by Clockwork-Muse** [comment]
@jhaygood86 - I dunno what they might be referencing, maybe it's one of the more niche messaging options?

I remember hearing a while ago that historically it was difficult to get .NET and Java to work together for SOAP, but that may no longer be the case.

**Comment by coxp** [comment]
@jhaygood86 Although not raised by myself, this seems to be the exact issue that we are suffering from. https://github.com/DigDes/SoapCore/issues/344.

We can't regenerate the client-side code because we don't own those services.

Additionally we have the issue that SoapCore is not really production ready (and to be fair is not v1 yet and even the author hopes that MS will make the project redundant). We built a service on SoapCore 0.9.9.1 and that version is not backwards compatible with newer versions and breaks existing clients if we upgrade to the latest version of SoapCore.

I don't want WCF. I just want MS to provide a way to create basic HTTP soap endpoints on the new stack. We're still creating new services as SOAP endpoints because that is what our clients require.
